lib/date_time/date_time.c

```diff
@@ -123,27 +206,152 @@ static int time_modem_get(void)
 		LOG_WRN("Did not get time from cellular network (error: %d). "
 			"This is normal as some cellular networks don't provide it or "
 			"time may not be available yet.", rc);
-		return rc;
+		return -ENODATA;
 	}
 
 	/* Relative to 1900, as per POSIX */
 	date_time.tm_year = date_time.tm_year + 2000 - 1900;
 	/* Range is 0-11, as per POSIX */
 	date_time.tm_mon = date_time.tm_mon - 1;
 
-	if (date_time.tm_year == MODEM_TIME_DEFAULT) {
-		LOG_DBG("Modem time never set");
-		return -ENODATA;
-	}
-
 	time_aux.date_time_utc = (int64_t)timeutil_timegm64(&date_time) * 1000;
 	time_aux.last_date_time_update = k_uptime_get();
 
 	LOG_DBG("Time obtained from cellular network");
 
 	return 0;
 }
+
+/**
+ * @brief Converts an octet having two semi-octets into a decimal.
+ *
+ * @details Semi-octet representation is explained in 3GPP TS 23.040 Section 9.1.2.3.
+ * An octet has semi-octets in the following order:
+ *   semi-octet-digit2, semi-octet-digit1
+ * Octet for decimal number '21' is hence represented as semi-octet bits:
+ *   00010010
+ *
+ * @param[in] value Octet to be converted.
+ *
+ * @return Decimal value.
+ */
+static uint8_t semioctet_to_dec(uint8_t value)
+{
+	/* 4 LSBs represent decimal that should be multiplied by 10. */
+	/* 4 MSBs represent decimal that should be added as is. */
+	return ((value & 0xf0) >> 4) + ((value & 0x0f) * 10);
+}
+
+static void date_time_at_xtime_handler(const char *notif)
+{
+	struct tm date_time;
+	uint8_t time_buf[6];
+	size_t time_buf_len;
+	char *time_str_start;
+	int err;
+
+	if (notif == NULL) {
+		return;
+	}
+	modem_valid_network_time = true;
```

Comment:
is this safe to be set already here? what if parsing fails?

Reply:
We've still received XTIME notification which indicates that modem has valid time. The rest of this file works more logically if we set it here instead of after the parsing (which I actually did originally)